narwhals/dataframe.py

```diff
@@ -1536,18 +1537,27 @@ def group_by(
             2  b  3  2
             3  c  3  1
         """
-        from narwhals.expr import Expr
         from narwhals.group_by import GroupBy
-        from narwhals.series import Series
 
         flat_keys = flatten(keys)
-        if any(isinstance(x, (Expr, Series)) for x in flat_keys):
-            msg = (
-                "`group_by` with expression or Series keys is not (yet?) supported.\n\n"
-                "Hint: instead of `df.group_by(nw.col('a'))`, use `df.group_by('a')`."
-            )
-            raise NotImplementedError(msg)
-        return GroupBy(self, *flat_keys, drop_null_keys=drop_null_keys)
+
+        if all(isinstance(key, str) for key in flat_keys):
+            return GroupBy(self, flat_keys, drop_null_keys=drop_null_keys)
+
+        from narwhals import col
+        from narwhals.expr import Expr
+        from narwhals.series import Series
+
+        _keys = [k if isinstance(k, (Expr, Series)) else col(k) for k in flat_keys]
+        expr_flat_keys, kinds = self._flatten_and_extract(*_keys)
```

Comment:
I've made this change locally and it only caused a single test to fail.

```suggestion
        expr_flat_keys, kinds = self._flatten_and_extract(*flat_keys)
```

I think the error message is reasonable and this is in `main` so I thought the plan was to raise for this anyway?

Did I miss/forget something? :dizzy_face:

```
E       narwhals.exceptions.InvalidIntoExprError: Expected an object which can be converted into an expression, got <class 'bool'>
E       
E       Hint:
E       - if you were trying to select a column which does not have a string
E         column name, then you should explicitly use `nw.col`.
E         For example, `df.select(nw.col(0))` if you have a column named `0`.
E       - if you were trying to create a new literal column, then you 
E         should explicitly use `nw.lit`.
E         For example, `df.select(nw.lit(0))` if you want to create a new
E         column with literal value `0`.

narwhals\dataframe.py:437: InvalidIntoExprError
```

Reply:
> I've made this change locally and it only caused a single test to fail.

This was to handle pandas column names that can be bool and int

Reply:
If were starting from scratch I think we could force users to wrap non string column names in nw.col , but seeing as we've allowed non string column names as arguments to group_by so far, we need to keep allowing that

Reply:
No worries folks